File: Day9_Minimize the Heights II/minimize_height.py
# https://www.geeksforgeeks.org/batch/gfg-160-problems/track/arrays-gfg-160/problem/minimize-the-heights3351
def getMinDiff(arr,k):
    # code here
    n = len(arr)
    arr.sort()
    small = arr[0]+k
    large = arr[n-1]-k
    result = arr[n-1]-arr[0]
    
    for i in range(n-1):
        mini = min(small,arr[i+1]-k)
        maxi = max(large, arr[i]+k)
        if mini < 0:
            continue
        result = min(result,maxi-mini)
    return result

user_k = int(input("Enter the value of k: "))
user_input = input("Enter space-separated integers: ")
arr = list(map(int, user_input.split()))
print(getMinDiff(arr=arr,k=user_k))

File: Day9_Minimize the Heights II/minimize_height.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class minimize_height {
    public static int getMinDiff(ArrayList<Integer> arr, int k) {
        Collections.sort(arr);
        int small = arr.getFirst()+k;
        int large = arr.getLast()-k;
        int result = arr.getLast() - arr.getFirst();
        for (int i = 0; i < arr.size()-1; i++) {
            int mini = Math.min(small, arr.get(i+1)-k);
            int maxi = Math.max(large, arr.get(i)+k);
            if(mini < 0)
                continue;
            result = Math.min(result,maxi-mini);
        }
        return result;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size of k: ");
        int k = sc.nextInt();
        System.out.println("Enter the size and element of array: ");
        int n = sc.nextInt();
        ArrayList<Integer> arr = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            arr.add(sc.nextInt());
        }
        int result = getMinDiff(arr, k);
        System.out.println("Minimum Height: "+ result);
        sc.close();
    }
}
